subsys/bluetooth/mesh/dfw.c

```diff
@@ -0,0 +1,5398 @@
+}
+
+int bt_mesh_dfw_relay_set(uint16_t net_idx, enum bt_mesh_feat_state state)
+{
+	return dfw_feature_set(net_idx, BT_MESH_DFW_RELAY_ENABLED, state);
+}
+
+int bt_mesh_dfw_proxy_get(uint16_t net_idx, enum bt_mesh_feat_state *state)
+{
+	return BT_MESH_FEATURE_NOT_SUPPORTED;
+}
+
+int bt_mesh_dfw_friend_get(uint16_t net_idx, enum bt_mesh_feat_state *state)
+{
+	if (!IS_ENABLED(CONFIG_BT_MESH_DFW_FRIEND)) {
+		*state = BT_MESH_FEATURE_NOT_SUPPORTED;
+		return 0;
+	}
+
+	return dfw_feature_get(net_idx, BT_MESH_DFW_FRIEND_ENABLED, state);
+}
+
+int bt_mesh_dfw_friend_set(uint16_t net_idx, enum bt_mesh_feat_state state)
+{
+	struct bt_mesh_subnet *sub = bt_mesh_subnet_get(net_idx);
+
+	if (!sub) {
+		LOG_ERR("Unable get subnet for network index 0x%04x", net_idx);
+		return -ENOENT;
+	}
+
+	if (!IS_ENABLED(CONFIG_BT_MESH_DFW_FRIEND)) {
+		return -ENOTSUP;
+	}
+
+	/* When the Friend state is set to 0x00, and directed friend functionality
+	 * is supported, then the Directed Friend state for all subnets shall be set
+	 * to 0x00 and shall not be changed.
+	 */
+	if ((bt_mesh_friend_get() == BT_MESH_FEATURE_DISABLED) &&
+	    (state == BT_MESH_FEATURE_ENABLED)) {
+		return -EACCES;
+	}
+
+	if (state != BT_MESH_FEATURE_DISABLED) {
+		return dfw_feature_set_by_subnet(sub, BT_MESH_DFW_FRIEND_ENABLED, state);
+	}
+
+	for (int i = 0; i < ARRAY_SIZE(bt_mesh.frnd); i++) {
+		struct bt_mesh_friend *frnd = &bt_mesh.frnd[i];
+		struct bt_mesh_dfw_node dependent;
+
+		if (!frnd->subnet || !frnd->established ||
+		    frnd->subnet->net_idx != net_idx) {
+			continue;
+		}
+
+		dependent.addr = frnd->lpn;
+		dependent.secondary_count = frnd->num_elem - 1;
+
+		(void)bt_mesh_dfw_dependent_node_update_start(sub->net_idx, &dependent, false);
+	}
+
+
+	return dfw_feature_set_by_subnet(sub, BT_MESH_DFW_FRIEND_ENABLED, state);
+}
+
+static inline bool dfw_is_dependent_node_enable(uint16_t net_idx)
+{
+	enum bt_mesh_feat_state friend_state = BT_MESH_FEATURE_DISABLED;
+	enum bt_mesh_feat_state proxy_state = BT_MESH_FEATURE_DISABLED;
+
+	(void)bt_mesh_dfw_friend_get(net_idx, &friend_state);
+	(void)bt_mesh_dfw_proxy_get(net_idx, &proxy_state);
+
+	return (friend_state == BT_MESH_FEATURE_ENABLED ||
+		proxy_state == BT_MESH_FEATURE_ENABLED);
+}
+
+static void subnet_evt(struct bt_mesh_subnet *sub, enum bt_mesh_key_evt evt)
+{
+	struct bt_mesh_dfw_subnet *dfw = NULL;
+	uint16_t net_idx;
+
+	switch (evt) {
+	case BT_MESH_KEY_ADDED:
+		for (int i = 0; i < ARRAY_SIZE(dfw_subnets); i++) {
+			net_idx = dfw_subnets[i].net_idx;
+			if (net_idx == BT_MESH_KEY_UNUSED) {
+				dfw = &dfw_subnets[i];
+				continue;
+			}
+
+			if (net_idx == sub->net_idx) {
+				sub->dfw = &dfw_subnets[i];
+				sub->dfw->net_idx = sub->net_idx;
+				return;
+			}
+		}
+
+		/* There is a one-to-one correspondence between dfw and subnet, and do not expect
+		 * that there will be no empty dfw for given subnet.
+		 */
+		if (!dfw) {
+			__ASSERT_NO_MSG(false);
+		}
+
+		sub->dfw = dfw;
```

Comment:
> 1. Could be all states encapsulated within the subnetwork structure?

@alxelax I need to explain to you why DFW states is not placed in `struct bt_mesh_subnet`, this is because during the process of calling `settings_load()`, we cannot guarantee the order in which the information of `netkey` and `dfw` will be restored.

Mesh Protocol 1.1 requires that the directed forwarding status of fixed paths and subnets needs to be stored. 

If you look at my implementation code, you can see that its corresponding storage path is `bt/mesh/DFW/Sub/Cfg/xxx`,  `bt/mesh/DFW/Sub/Fw/xxx/xxx/xxx` and `bt/mesh/DFW/Sub/Dep/xxx/xxx/xxx/xxx`.

According to my implementation, when `settings_load()` is executed, if the dfw information is restored first, only `net_idx` will be temporarily stored in the `struct bt_mesh_dfw_subnet`. After waiting for the subnet to be restored, a **one-to-one** correspondence will be established between the `subnet` and `dfw` in `BT_MESH_KEY_ADDED` .